### ISSUE TITLE
Conditional code preprocessing

### ISSUE THREAD

**Issue by dnfield** [dart-lang/sdk#33249]
I'm primarily concerned about this for Flutter, where I can't use `dart:mirrors`.

As a library author, I would like to be able to publish code that conditionally uses new functionality if it's available in the consumer's SDK version.

I could imagine implementing this by checking:

1. Whether a compile time variable is set.
2. Whether a method/property is defined/exists in the Flutter SDK API that my library consumes.

For example, using C-style preprocessor notation

```
#if FLUTTER_SDK >= 0.4.5
// Use new and better methods, use alternatives to deprecated methods, e.g. `utf8.decode`
#else
// Fall back or simply don't provide functionality, or use a method
// that isn't deprecated in older version, e.g. `UTF8.decode`
#endif
```

or:

```
#if IS_DEFINED(sdkClass.newCoolMethod)
// code that uses `sdkClass.newCoolMethod()
#else
// code that uses some work around, or perhaps does nothing at all
#endif
```

Basically, I'd like to avoid needing to maintain multiple versions of my library to support Flutter beta, dev, and/or master channels.  I'm OK with users on Beta not getting cutting edge functionality, but I'd like to avoid a very confusing scenario where I have to keep going back and checking if I can migrate functionality to my "beta" tracking package from my "dev" tracking package - just let the compiler include that functionality if it's available in the SDK, and skip it (or use a specified alternative) if it's not.

Although I'm using C-style preprocessing to illustrate here, I'm not looking for a full preprocessor.  I'd be very open to other ideas or suggestions that don't involve reflection at runtime.

**Comment by robertmuth** [comment]
some mechanism like this would also be useful for libraries that 
come in two flavors: one depending on 'dart:html' and one that does not.

**Comment by leafpetersen** [comment]
In the Dart 2 time frame, we will have some (but not all of what you are asking for here).  You will be able to conditionally import/export code controlled by a fixed set of platform defined flags.  We will not have finer granularity (e.g. platform version) in that time frame (and no current plans for adding it later). Follow https://github.com/dart-lang/sdk/issues/32960 for more details.

**Comment by dnfield** [comment]
So it certainly could be achieved as a preprocessing step before passing to Dart if it's not targetted for the language.  My only fear with that is fragmentation (there's the coffeedart preprocessor, and the typedart preprocessor, and the danfielddart preprocessor, etc).  That, and the fact that it wouldn't be recognized by the analyzer in any meaningful way.

**Comment by lrhn** [comment]
I do want a feature like this. It enables gradual migration of new, breaking, language features.

**Comment by robertmuth** [comment]
BTW: this is the best documentation on conditional imports I could find:

https://medium.com/@dvargahali/dart-2-conditional-imports-update-16147a776aa8

**Comment by dnfield** [comment]
This is coming up as something that would help Flutter framework implement cross-platform features in ways that are especially challenging today - such as supporting both `dart:html` and `dart:io`, or supporting Fuchsia specific things that we don't expect will be available for other platforms.

**Comment by robertmuth** [comment]
After having used conditional imports on a few occasions I find them to be rather clumsy.
Typically I use them to redirect code to a stub implementation, which usually affects many files.

A much cleaner mechanism in my opinion would allow me to specify this redirection in only  place.
For example, I would like to say 
"for this build target, redirect all uses of dart:html to  stubbed_out_html.dart"

**Comment by lrhn** [comment]
If you just want to implement an entire API twice, without forwarding from stubs, you can use conditional *exports*.

```dart
library platform_dependent_html;
export "stubbed_out_html.dart"
    if (dart.library.html) "html.dart";
```

Then you can import this library, and get real HTML if it's there, or `stubbed_out_html.dart` if not.

We do not provide a way for a build to replace an existing library that someone else has asked for explicitly.

**Comment by natebosch** [comment]
> If you just want to implement an entire API twice, without forwarding from stubs, you can use conditional exports.

I don't think conditional exports are working everywhere currently. #34179

**Comment by DQvsRA** [comment]
Much needed feature! 
And you can find inspiration in Haxe Conditional Compilation it helps us a lot https://haxe.org/manual/lf-condition-compilation.html

**Comment by maxim-saplin** [comment]
I'm definetly missing C# pre-processor features, different build configurations defining own symbols and condition complications with #if

**Comment by navaronbracke** [comment]
@lrhn has there been any more thought about this recently? I was thinking about this exact proposal in lieu of having a much cleaner way of doing io/html implementations of let's say a service class (with regards to Flutter).

Currently what I have to do to get it working is the following.
Not sure if there is a cleaner way for my current approach either.
(Bear in mind, this is in a standalone project, *specifically not* a package)

1. A stub file with a global method
  - contains a top-level method like `MyService getMyServiceInstance() => throw UnimplementedError('this is a stub');`
2. A common interface that defines the service, MyService
  - contains a conditional import that imports the correct implementation (or the stub)
  - contains a factory constructor of `MyService` that uses the global method as implementation
3. Implementations for io/web
  - override the global stub method and pass in their own instance

With pre-processor features like in C++/C# I can get rid of the stub file and the separate implementations, therefore keeping everything in one class. 

Something like this

```
#ifdef library.dart.io
import 'dart:io';
#elifdef library.dart.html
import 'dart:html';
#endif

class MyService {
  bool someFunction(){
#ifdef library.dart.io
    return true;
#elifdef library.dart.html
    return false;
#else
    throw UnimplementedError();
#endif
  }
}
```

**Comment by martin-braun** [comment]
I have to live with a warning now, because I try to support my package for Flutter 2 and 3 and a null-aware operator can be avoided in version 3. If I had a compiler directive that allows me to evaluate the major Flutter version, I could avoid the warning. Ref: https://github.com/martin-braun/google_maps_place_picker_mb/commit/83357f845f13fcb0cb0fbf0bfc1b8ad177b0d214#commitcomment-75456126

**Comment by fuzzybinary** [comment]
I'd like to know if the Dart team is at least considering this, especially for package maintainers, as we can't assume people using our packages will upgrade to the most recent version of Flutter or Dart.

A few things I have encountered recently that could be easily solved using conditional compilation, but require workarounds (or worse, maintaining multiple versions of the same library).

* The addition of two methods to HttpClient in Dart 2.17 (required I ship two versions of my package, one that supports < 2.16 and one >= 2.17).
* The change to make WidgetsBinding.instance non-optional 
* Upcoming changes to prefer using PlatformDispatcher.instance.onError over runZonedGuarded

**Comment by dnfield** [comment]
The more I think about this, the more I think what I'd really like here is API availability capabilities.

IOW, it's not that I really want preprocessor defines in Dart, it's that I want to know if some API is available (e.g. whether dart:ui as a whole is avialable, or whether dart:ui has a function called someNewAwesomeThing, or whether object X has a const constructor not, etc.).

**Comment by martin-braun** [comment]
> * The change to make WidgetsBinding.instance non-optional

I figured out a way to circumvent this issue. 

```dart
/// This allows a value of type T or T? to be treated as a value of type T?.
///
/// We use this so that APIs that have become non-nullable can still be used
/// with `!` and `?` on the stable branch.
// TODO: Remove this, once Flutter 2 support is dropped
// See https://github.com/flutter/flutter/issues/64830
T? _ambiguate<T>(T? value) => value;
```

Now you can do `_ambiguate(WidgetsBinding.instance)!.addPostFrameCallback(...`.

**Comment by VladimirCores** [comment]
In Haxe language conditional compilation looks like this:
```haxe
class Main {
  public static function main() {
    #if !debug
    trace("ok");
    #elseif (debug_level > 3)
    trace(3);
    #else
    trace("debug level too low");
    #end
  }
}
```
https://haxe.org/manual/lf-condition-compilation.html

**Comment by ohir** [comment]
@dnfield 
> it's not that I really want preprocessor defines in Dart, it's that I want to know if some API is available (e.g. whether dart:ui as a whole is avialable, or whether dart:ui has a function called someNewAwesomeThing, or whether object X has a const constructor not, etc.).

But by what mechanics would you act on this knowledge? At the runtime all mentioned pieces can be known just by a generated consts – as Flutter installation comes with Dart sources.

**Comment by dnfield** [comment]
@ohir - there's no sane way currently to check at runtime whether a particular method (or method signature) is defined, and certain things are checked at compile time (like whether a const constructor is available or not).

I'm mainly thinking of how availability macros work in Objective-C on iOS/macOS. There I can do things like have code conditionally work if and only if I'm using a particular SDK version.

So for example, as a package author, I want to use some new exciting API that Flutter offers, but I have no way to check whether that API exists today (ignoring doing things like casting to dynamic and trying to invoke a method which is horrible for many reasons). I also don't have a way to have a single package offer better behavior depending on which version of the Flutter/Dart SDKs are in use. E.g.

```dart
  if (DartVersion > xxx) {
   // use cool new feature that is faster
  } else {
    // use slower fallback
  }
```

You can come close to getting this by having minimum version requirements in your pubspec, but pub has a hard time solving for strange and overlapping constraints - and it ends up requiring package maintainers to maintain multiple versions of their packages that may not even make it to their users correctly because of pub.

**Comment by fuzzybinary** [comment]
I asked @munificent about this on Twitter and he mentioned that a general case pre-processor likely isn't coming any time soon, because it requires a lot of the tooling support for things like changing targets (think the target / platform selector on Visual Studio), and since we have `Platform.isX`, `kDebugMode`, `kIsWeb` and a few other compile time consts using `Platform.environment`, the one thing library creators / maintaners are lacking is what @dnfield is describing.

I feel like Dart could copy swift's `available` attributes and compiler directive for use with Dart and Flutter versions: https://www.avanderlee.com/swift/available-deprecated-renamed/

So similar to what Dan is suggesting we could have code like:
```
class MyHttpClient : HttpClient {
  @available(Dart >= '2.17')
  @override
  set connectionFactory(
          Future<ConnectionTask<Socket>> Function(
                  Uri url, String? proxyHost, int? proxyPort)?
              f) =>
      innerClient.connectionFactory = f;
}
```

Or:
```
if (#available(Flutter > 3)) {
  WidgetsBinding.instance.addObserver(this)
} else {
   WidgetsBinding?.instance.addObserver(this)
}
```

In both cases, the compiler should skip any code that doesn't meet the available criteria.

Since this mostly affects library maintainers, and we're likely to use fvm to change versions of Flutter, any support needed in tooling other than the analyzer could lag.

I'd be interested to hear what the Dart team thinks...

**Comment by ohir** [comment]
> how availability macros work in Objective-C on iOS/macOS

Similar functionality likely could be provided for Dart/Flutter own inventory, but for a substantial effort. I have no high hopes this will ever happen (ie. maintaining full lifecycles _and_ supporting them thorough the toolchain).

> pub has a hard time solving for strange and overlapping constraints

What about library author testing for all possible permutations (of execution path configurations) that are result of the pub solver coming to a given set of dependencies peculiar to the site of use? You can control, and that just to an extent, only your package's direct dependencies (version-wise).

> I have no way to check whether that API exists today

I understand that you have no way to check whether that API exists at your package user's installation, am I correct? If I am, I know of no "contemporary automagic solution", but I think that 'ol pale `./configure` might work (if contemporary users were used to do eg. a `dart test configure` step). There, in a first test, could be a place to do that ugly _cast-try-catch_ research for API availability. Done once.

**Comment by q25001560** [comment]
Members! please add the conditional compilation feature to Dart just like C or Apple's Swift. It's important!

**Comment by BradKwon** [comment]
I am also missing C# pre-processor directives. Hope Dart will have this feature soon as well.

**Comment by esDotDev** [comment]
We're missing this over here as well: 
https://github.com/gskinner/flutter_animate/issues/41

This is a bit downstream, but it would be quite nice to have the flutter versions eventually stored off as compilation consts:
```c#
#if flutter_3_3_1_or_newer
 //
#elif flutter_3_3_0
//
#else
 //
#endif 
```

So it would be cool if somehow that usecase was enabled (not just dart version).

**Comment by martin-braun** [comment]
It would be nice to have `kDebugMode` as pre-processor variable as well. It would also be nice to have this ability in the pubspec.yaml somehow, or fix `dev_dependencies` already.

`dev_dependencies` apparently are shipped into the release build of Android (when imported in code, despite being used behind `kDebugMode` exclusively).

**Comment by mraleph** [comment]
> `dev_dependencies` apparently are shipped into the release build of Android (when imported in code, despite being used behind `kDebugMode` exclusively).

If all uses are hidden behind `if (kDebugMode)` I would expect Dart code to be tree-shaken. Native dependencies will still be pulled in though.

Do you have an example?

**Comment by martin-braun** [comment]
> Native dependencies will still be pulled in though.

Yes, exactly this is my problem. Please see the backgrounds for my complaint here: https://github.com/Baseflow/flutter-geolocator/issues/1296

The package in question that causes this problem though is `safe_device`, see https://github.com/ufukhawk/safe_device/issues/33

I only ever need the package in debug mode, so I assumed defining it in `dev_dependencies` would strip the native bits as well. Apparently, the package causes a lot of problems in the Android release build that are massive and critical, but I still rely on the package during development, it makes things so much easier.

**Comment by andynewman10** [comment]
> If all uses are hidden behind `if (kDebugMode)` I would expect Dart code to be tree-shaken. Native dependencies will still be pulled in though.

Why will native `dev_dependencies` be pulled? `dev_dependencies` assets are also shipped in the release build, which seems also wrong to me.

I see two important-to-fix bugs here, as there is no easy way for developers to work on development features and have the possibility to easily deactivate these features for release builds.

**Comment by yaminet1024** [comment]
I also need this to distinguish between different flutter sdk versions

**Comment by munificent** [comment]
No update that I'm aware of.

**Comment by fuzzybinary** [comment]
@munificent Is this something macros could be hacked a bit to do?

For example, could you write:

```dart
  @dart_minimum('2.17')
  @override
  set connectionFactory(
          Future<ConnectionTask<Socket>> Function(
                  Uri url, String? proxyHost, int? proxyPort)?
              f) =>
      innerClient.connectionFactory = f;
```

And have the macro read current Dart version and remove the attributed code block for versions under 2.17? Or are macros only additive?

This is not ideal (It'd still love a `#available` style system) but might be a good temporary workaround.

**Comment by lrhn** [comment]
Macros are *mainly* additive, but they can bypass existing code entirely by not calling it.
An augmentation for the setter could be
```dart
augment set connectionFactory(Future<ConnectionTask<Socket>> Function(
                  Uri url, String? proxyHost, int? proxyPort)?
              f) {}
```
or maybe a body of `{ super.connectionFactory = f; }` if there is a super-implementation to fall back on.

The function won't stop existing, but it will stop doing anything.

A bigger questions is whether the macro has access to the language version (but I guess you could just have one version of the macro package per SDK release, say one with an SDK constraint of `>= 3.7.0 < 3.8.0` which knows that it's definitely running on SDK 3.7.X and language version 3.7.
(Maybe we should just make the information available, so we don't get people filling up the package version database like that.)

**Comment by munificent** [comment]
> @munificent Is this something macros could be hacked a bit to do?

Not really. By design, macros don't have access to any configuration information. If they did, then the compilers would have to be very careful to run macros separately for each possible configuration and track their outputs separately.

It would definitely be useful of macros could do config-specific stuff, but it has a *lot* of implications for developer experience, complexity, and compile time performance, so at least for now we are avoiding it by not giving macros access to that information.

cc @jakemac53
